pkgsCross.wasi32.buildPackages.rustc: fix build

diogotcorreia opened the pull request:
Fixes #463589

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

diogotcorreia committed:
pkgsCross.wasi32.buildPackages.rustc: fix build

Fixes #463589

PerchunPak commented on pkgs/development/compilers/rust/rustc.nix:
Thank you for this PR! I want to package some wasi Rust packages, but I cannot make this PR work. Can you please provide a minimal example with buildRustPackage?

philiptaron reviewed (approved):
I validated that `pkgsCross.wasi32.buildPackages.rustc` is broken on master and builds after this.

PerchunPak reviewed (changes requested):
Currently, the packages need to provide these three lines to work

```nix
cargoBuildFlags = [ "--target wasm32-wasip1" ];
env.RUSTFLAGS = "-C linker=wasm-ld";
nativeBuildInputs = [ buildPackages.lld ];
```

It is definitely a blocker. But I successfully packaged a dozen plugins, so the code works.

alyssais commented:
I agree, but that's not a reason not to fix the rustc build.  It should be fixed, but I don't see what it has to do with this change.

rvolosatovs commented on pkgs/development/compilers/rust/rustc.nix:
```diff
@@ -250,6 +251,17 @@ stdenv.mkDerivation (finalAttrs: {
+      let
+        wasiSysRoot = runCommand "wasi-sysroot" { } ''
+          mkdir -p $out/lib
+          ln -s ${pkgsBuildTarget.targetPackages.stdenv.cc.libc}/lib $out/lib/wasm32-wasip1
```
`wasm32-wasi` has been renamed to `wasm32-wasip1`, in some tools (like LLVM) `wasm32-wasi` is an *alias* for `wasm32-wasip1`. They should be equivalent.

`wasm32-wasip2` is the more modern target, while `wasm32-wasip3` is still in-development.

I'm not actually sure which version of wasi `nixpkgs` targets by default, but I'd guess it's `wasm32-wasip1`.

I would suggest setting `TARGET_TRIPLE=wasm32-wasip1` explicitly

diogotcorreia replied:
I've explicitly set `TARGET_TRIPLE` to `stdenv.system` (`wasm32-wasi`), since that is what version 27 of wasilibc is using: https://github.com/WebAssembly/wasi-libc/blob/wasi-sdk-27/Makefile#L50

In version 31 (current latest), `wasm32-wasip1` is the default instead (even though they are equivalent), so whenever we update to that version we can revisit this. https://github.com/WebAssembly/wasi-libc/blob/wasi-sdk-31/CMakeLists.txt#L39
They are now using CMake btw, so the derivation will need quite a few changes.
Also keep in mind that `TARGET_TRIPLE` is version 27 does not actually allow building `wasm-wasip2` if we wanted, we'd need to set `WASI_SNAPSHOT` for that. This is not an issue in version 31 anymore.

I am now creating symlinks `lib/wasm32-wasip1` and `lib/wasm32-wasi` pointing to `lib`. Not sure if this "cycle" will cause issues.

PerchunPak committed:
doc: add wasip1 rust package example

PerchunPak commented:
I won't oppose merging this, but please apply [this patch](https://github.com/NixOS/nixpkgs/commit/41c794adc90ff536a6bd8fa5905939778b055119) to your commit first. Feel free to add more sentences to the docs, I don't know what to put there

alyssais commented on pkgs/by-name/wa/wasilibc/package.nix:
```diff
@@ -47,6 +47,7 @@ stdenvNoLibc.mkDerivation (finalAttrs: {
       "SYSROOT_LIB:=$SYSROOT_LIB"
       "SYSROOT_INC:=$SYSROOT_INC"
       "SYSROOT_SHARE:=$SYSROOT_SHARE"
+      "TARGET_TRIPLE:=${stdenvNoLibc.system}"
```
This probably ought to be a separate commit, given it's not necessary to fix the rustc build, right?

diogotcorreia replied:
Sure, I can split it into two commits

diogotcorreia committed:
wasilibc: explicitly set TARGET_TRIPLE

While this is no-op for wasm32-wasip1, this change future-proofs
wasilibc for when nixpkgs supports wasm32-wasip2 and wasm32-wasip3.

PerchunPak reviewed (approved):
After the recent changes, builds without `cargoBuildFlags = [ "--target wasm32-wasip1" ];` pass for me too and the package works normal

alyssais commented on pkgs/development/compilers/rust/rustc.nix:
```diff
@@ -11,6 +11,7 @@
   llvmSharedForHost,
   llvmSharedForTarget,
   llvmPackages,
+  runCommand,
```
This is not used.

diogotcorreia replied:
Thanks, leftover from the previous iteration of this PR. Removed.

diogotcorreia reviewed (commented):
Partial run of nixpkgs-review since my PC can't build firefox without dying 😅 

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 463720 --package pkgsCross.wasi32.libc --package pkgsCross.wasi32.buildPackages.rustc`
Commit: `72b5a1c0e616b4c440935b7e6843765a8411bb86`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 6 packages built:</summary>
  <ul>
    <li>pkgsCross.wasi32.buildPackages.rustc</li>
    <li>pkgsCross.wasi32.buildPackages.rustc.doc (pkgsCross.wasi32.buildPackages.rustc.doc.doc, pkgsCross.wasi32.buildPackages.rustc.doc.man)</li>
    <li>pkgsCross.wasi32.buildPackages.rustc.man (pkgsCross.wasi32.buildPackages.rustc.man.doc, pkgsCross.wasi32.buildPackages.rustc.man.man)</li>
    <li>pkgsCross.wasi32.libc</li>
    <li>pkgsCross.wasi32.libc.dev (pkgsCross.wasi32.libc.dev.dev, pkgsCross.wasi32.libc.dev.share)</li>
    <li>pkgsCross.wasi32.libc.share (pkgsCross.wasi32.libc.share.dev, pkgsCross.wasi32.libc.share.share)</li>
  </ul>
</details>
